versioned_hdf5/tests/test_staged_changes.py

```diff
@@ -0,0 +1,705 @@
+
+    assert a.dtype == "f4"
+    for slab in a.slabs:
+        assert slab is None or slab.dtype == "f4"
+
+    assert b.dtype == "i2"
+    for slab in b.slabs:
+        assert slab is None or slab.dtype == "i2"
+
+    assert_array_equal(a, np.array([[1, 4, 0]], dtype="f4"), strict=True)
+    assert_array_equal(b, np.array([[1, 4, 0]], dtype="i2"), strict=True)
+
+
+def test_refill():
+    # Actual base slabs can entirely or partially contain the fill_value
+    a = StagedChangesArray.from_array([[1], [42]], chunk_size=(2, 1), fill_value=42)
+    a.resize((2, 3))  # Create full chunks full of 42
+    a[0, 2] = 2  # Create staged slab with a non-42 and a 42
+    assert_array_equal(a, [[1, 42, 2], [42, 42, 42]])
+    assert_array_equal(a.slab_indices, [[1, 0, 2]])
+
+    b = a.refill(99)
+    # a is unchanged
+    assert_array_equal(a, [[1, 42, 2], [42, 42, 42]])
+    assert_array_equal(a.slab_indices, [[1, 0, 2]])
+
+    # full slabs -> still full slabs, but slabs[0] has changed value
+    # base slabs -> staged slabs with 42 points replaced with 99
+    # staged slabs -> same slab index, but 42 points have been replaced with 99
+    assert_array_equal(b, [[1, 99, 2], [99, 99, 99]])
+    assert_array_equal(b.slab_indices, [[3, 0, 2]])
+
+
+def test_big_O_performance():
+    """Test that __getitem__ and __setitem__ performance is
+    O(selected number of chunks) and not O(total number of chunks).
+
+    Note: resize() is, by necessity, O(total number of chunks in the array) when
+    adding/removing chunks and O(total number of chunks along resized axis) when
+    enlarging the edge chunks without changing the number of chunks.
+    """
+
+    def benchmark(shape):
+        arr = StagedChangesArray.full(shape, (1, 2))
+        # Don't measure page faults on first access to slab_indices and slab_offsets.
+        # In the 10 chunks use case, it's almost certainly reused memory.
+        _ = arr[0, -3:]
+
+        t0 = time.thread_time()
+        # Let's access the end, just in case there's something that
+        # performs a full scan which stops when the selection ends.
+
+        # Update only part of a chunk. This triggers a whole
+        # extra section worth of logic in __setitem__.
+        arr[0, -1] = 42
+        assert arr[0, -1] == 42
+        assert arr[0, -2] == 0
+        t1 = time.thread_time()
+        return t1 - t0
+
+    # trivially sized baseline: 5 chunks
+    a = benchmark((1, 10))
+
+    # 5 million chunks, small rulers
+    # Test will trip if __getitem__ or __setitem__ perform a
+    # full scan of slab_indices and/or slab_offsets anywhere
+    b = benchmark((2_500, 4_000))
+    np.testing.assert_allclose(b, a, rtol=0.2)
+
+    # 5 million chunks, long rulers
+    # Test will trip if __getitem__ or __setitem__ construct or iterate upon a ruler as
+    # long as the number of chunks along one axis, e.g. np.arange(mapper.n_chunks)
+    c = benchmark((1, 10_000_000))
+    np.testing.assert_allclose(c, a, rtol=0.2)
```

Comment:
How reliable do you expect this to be in CI? If this is flaky I'm guessing there's a way we can measure this with `Mock`, though I suppose there could be something on the h5py side of things that we wouldn't be able to control :thinking:

Reply:
Very reliable.
Even if the hypervisor stole away the CPU while running either use case, which would arbitrarily prolong the wall time, thread time would not increase.

Reply:
None of test_staged_changes.py or staged_changes.py uses h5py. It's all plain numpy.